src/vs/workbench/contrib/notebook/browser/viewModel/notebookOutlineProviderFactory.ts

```diff
@@ -0,0 +1,39 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ReferenceCollection, type IReference } from 'vs/base/common/lifecycle';
+import { IInstantiationService, createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import type { INotebookEditor } from 'vs/workbench/contrib/notebook/browser/notebookBrowser';
+import { NotebookCellOutlineProvider } from 'vs/workbench/contrib/notebook/browser/viewModel/notebookOutlineProvider';
+import type { OutlineTarget } from 'vs/workbench/services/outline/browser/outline';
+
+class NotebookCellOutlineProviderReferenceCollection extends ReferenceCollection<NotebookCellOutlineProvider> {
+	constructor(@IInstantiationService private readonly instantiationService: IInstantiationService) {
+		super();
+	}
+	protected override createReferencedObject(_key: string, editor: INotebookEditor, target: OutlineTarget): NotebookCellOutlineProvider {
+		return this.instantiationService.createInstance(NotebookCellOutlineProvider, editor, target);
+	}
+	protected override destroyReferencedObject(_key: string, object: NotebookCellOutlineProvider): void {
+		object.dispose();
+	}
+}
+
+export const INotebookCellOutlineProviderFactory = createDecorator<INotebookCellOutlineProviderFactory>('INotebookCellOutlineProviderFactory');
+
+export interface INotebookCellOutlineProviderFactory {
+	getOrCreate(editor: INotebookEditor, target: OutlineTarget): IReference<NotebookCellOutlineProvider>
+}
+
+export class NotebookCellOutlineProviderFactory implements INotebookCellOutlineProviderFactory {
```

Comment:
3 outline providers are created.
Meaning each key stroke in cell editors results in outlines being calculated 3 times.
Assume we take 50ms to calculate it, then we spend 50+50+50ms in total for each keystroke

Now it will be only one calculation.

Reply:
Did you actually see the provider for the goto symbols being calculated on each keystroke? I thought that one was only called when you open that widget

Reply:
Open the widget, and every keystroke will result in outline being calculated.

Reply:
> only called when you open that widget

Given the fact taht this is also used by other widgets, its possible it will re-compute it,
E.g. when the Sticky scroll feature is eanbled, we re-compute on every key stroke even if the outline widget is not visible.